src/trait-bounds.md

```diff
@@ -0,0 +1,131 @@
+> [_ForLifetimes_](#higher-ranked-trait-bounds)<sup>?</sup> [_TraitPath_]  
+> &nbsp;&nbsp; | `(` `?`<sup>?</sup>
+> [_ForLifetimes_](#higher-ranked-trait-bounds)<sup>?</sup> [_TraitPath_] `)`  
+>  
+> _LifetimeBounds_ :  
+> &nbsp;&nbsp; `:` [_Lifetime_] ( `+` [_Lifetime_] )<sup>\*</sup> `+`<sup>?</sup>  
+
+[Trait] and lifetime bounds provide a way for [generic items][generic] to
+restrict which types and lifetimes are used as their parameters. Bounds can be
+provided on any type in a [where clause]. There are also shorter alternative
+forms:
+
+* Bounds written after declaring a [generic parameter][generic]:
+  `fn f<A: Copy>() {}` is the same as `fn f<A> where A: Copy () {}`.
+* In trait declarations as [supertraits]: `trait Circle : Shape {}` is
+  equivalent to `trait Circle where Self : Shape {}`.
+* In trait declarations as bounds on [associated types]:
+  `trait A { type B: Copy; }` is equivalent to
+  `trait A where Self::B: Copy { type B; }`.
+
+Bounds on an item must be satisfied when using the item. When type checking
+and borrow checking the item, the bounds can be used to prove that a type
+implements a required bound.
+
+```rust
+# type Surface = i32;
+trait Shape {
+    fn draw(&self, Surface);
+    fn name() -> &'static str;
+}
+
+fn draw_twice<T: Shape>(surface: Surface, sh: T) {
+    sh.draw(surface);           // Can call method because T: Shape
+    sh.draw(surface);
+}
+
+fn copy_and_draw_twice<T: Copy>(surface: Surface, sh: T) where T: Shape {
+    let shape_copy = sh;        // doesn't move sh because T: Copy
+    draw_twice(surface, sh);    // Can use generic function because T: Shape
+}
+
+struct Figure<S: Shape>(S, S);
+
+fn name_figure<U: Shape>(
+    figure: Figure<U>,          // Type Figure<U> is well-formed because U: Shape
+) {
+    println!(
+        "Figure of two {}",
+        U::name(),              // Can use associated function
+    );
+}
+```
+
+Trait and lifetime bounds are also used to name [trait objects].
+
+> [^sized_only] `?` is only used to declare that the [`Sized`] trait may not be
+> implemented for a type parameter or associated type.
+
+## Lifetime bounds
+
+Lifetime bounds can be applied to other lifetimes, or to types. The bound
+`'a: 'b` is usually read as `'a` *outlives* `'b`. `'a: 'b` means that `'a`
+includes all points in the code that `'b` does.
+
+```rust
+fn f<'a, 'b>(x: &'a i32, mut y: &'b i32) where 'a: 'b {
+    y = x;                      // &'a i32 is a subtype of &'b i32 because 'a: 'b
+    let r: &'b &'a i32 = &&0;   // &'b &'a i32 is well formed because 'a: 'b
+}
+```
+
+`T: 'a` means that all lifetime parameters of `T` outlive `'a`. For example if
+`'a` is an unconstrained lifetime parameter then `i32: 'static` and
+`&'static str: 'a` are satisfied but `Vec<&'a ()>: 'static` is not.
+
+## Higher-ranked trait bounds
+
+> **<sup>Syntax</sup>**  
+> _ForLifetimes_ :  
+> &nbsp;&nbsp; `for` `<` ( [_LifetimeParams_][generic][^unused_for_bounds]
```

Comment:
The same issue is happening here with the `[^unused_for_bounds]` thing. I think the grammar is busy enough without footnotes distracting you away from it.